core/src/main/java/org/neo4j/ogm/context/SingleUseEntityMapper.java

```diff
@@ -133,7 +133,12 @@ private void writeProperty(ClassInfo classInfo, Object instance, Map.Entry<Strin
         if (writer == null) {
             logger.warn("Unable to find property: {} on class: {} for writing", key, classInfo.name());
         } else {
-            Class elementType = DescriptorMappings.getType(writer.getTypeDescriptor());
+            // This takes attribute and composite converters into consideration.
+            Class<?> elementType = writer.convertedType();
+            if (elementType == null) {
+                // If it is not a converted type, we retrieve the element type (not the field type, which maybe a collection)
+                elementType = DescriptorMappings.getType(writer.getTypeDescriptor());
+            }
             boolean targetIsCollection = writer.type().isArray() || Iterable.class.isAssignableFrom(writer.type());
```

Comment:
`writer.type()` internally calls `writer.convertedType()` with the fallback to attribute fieldType. The logic in  `writer.convertedType()` is quite expensive and is now called at least three times, in case of a collection even a fourth time.

I suggest adding `getFieldType()` to `FieldInfo` and changing this code as follows:
```
// This takes attribute and composite converters into consideration.
Class<?> fieldType = writer.convertedType();
Class<?> elementType;
if (fieldType == null) {
    // If it is not a converted type, we retrieve the element type (not the field type, which maybe a collection)
    elementType = DescriptorMappings.getType(writer.getTypeDescriptor());
    fieldType = writer.getFieldType();
} else {
    // If it is a converted type, use it as is
    elementType = fieldType;
}
boolean targetIsCollection = fieldType.isArray() || Iterable.class.isAssignableFrom(fieldType);
```
Note: `fieldType` can then also be reused for calling `EntityAccessManager.merge(...)`.

I know that performance improvement is not necessarily part of this issue but neither should be performance degradation.

Reply:
I see the pain with this and I tried a couple of things on the `FieldInfo`, but tbh, that's like a house of cards around the field type. 

What I would like to see is something like `getDeclaredFieldType` and `getEffectiveGraphPropertyType` and only collection checking inside the field info itself. 

I tried a couple of things, but those caused so many test to fail, that I would rather not do this in a patch.

The changes with my polishing (see below) I'm fine with for a patch, the later i see that we find some time.